File: hangman.py
# hangman.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import os
import random
import json
from pathlib import Path

# ANSI-цвета
COLORS = {
    'reset': '\033[0m',
    'green': '\033[92m',
    'red': '\033[91m',
    'yellow': '\033[93m',
    'blue': '\033[94m',
    'bold': '\033[1m'
}

def colorize(text, color):
    return f"{COLORS.get(color, '')}{text}{COLORS['reset']}"

# Словари по темам
WORDS = {
    'animals': ['кот', 'собака', 'тигр', 'слон', 'жираф', 'дельфин', 'медведь', 'волк', 'лиса', 'заяц'],
    'cities': ['москва', 'париж', 'лондон', 'берлин', 'рим', 'токио', 'пекин', 'нью-йорк', 'каир', 'мехико'],
    'food': ['пицца', 'суши', 'бургер', 'паста', 'салат', 'суп', 'стейк', 'омлет', 'торт', 'мороженое'],
    'professions': ['врач', 'учитель', 'инженер', 'пилот', 'повар', 'художник', 'музыкант', 'архитектор', 'юрист', 'журналист'],
    'general': ['программирование', 'алгоритм', 'компьютер', 'интернет', 'игра', 'код', 'данные', 'сервер', 'сайт', 'приложение']
}

# ASCII-арт виселицы (11 стадий: 0 - пусто, 10 - полная)
HANGMAN_PICS = [
    """
      +---+
          |
          |
          |
          |
          |
    =========
    """,
    """
      +---+
      |   |
          |
          |
          |
          |
    =========
    """,
    """
      +---+
      |   |
      O   |
          |
          |
          |
    =========
    """,
    """
      +---+
      |   |
      O   |
      |   |
          |
          |
    =========
    """,
    """
      +---+
      |   |
      O   |
     /|   |
          |
          |
    =========
    """,
    """
      +---+
      |   |
      O   |
     /|\\  |
          |
          |
    =========
    """,
    """
      +---+
      |   |
      O   |
     /|\\  |
     /    |
          |
    =========
    """,
    """
      +---+
      |   |
      O   |
     /|\\  |
     / \\  |
          |
    =========
    """
]

class Hangman:
    def __init__(self, max_errors=10, theme='general'):
        self.max_errors = max_errors
        self.theme = theme
        self.word = random.choice(WORDS.get(theme, WORDS['general']))
        self.display = ['_'] * len(self.word)
        self.guessed = set()
        self.wrong = set()
        self.errors = 0
        self.hint_used = False
        self.game_over = False
        self.won = False
        self.stats_file = Path.home() / '.hangman_stats.json'
        self.load_stats()

    def load_stats(self):
        if self.stats_file.exists():
            with open(self.stats_file, 'r') as f:
                self.stats = json.load(f)
        else:
            self.stats = {'games': 0, 'wins': 0, 'losses': 0, 'streak': 0, 'best_streak': 0}

    def save_stats(self):
        with open(self.stats_file, 'w') as f:
            json.dump(self.stats, f, indent=2)

    def display(self):
        os.system('clear' if os.name == 'posix' else 'cls')
        print(colorize("🎯 ВИСЕЛИЦА", 'bold'))
        print(colorize(f"Тема: {self.theme}", 'blue'))
        print(colorize(f"Ошибок: {self.errors}/{self.max_errors}", 'red' if self.errors > 0 else 'green'))
        print(HANGMAN_PICS[self.errors])
        print("Слово: " + ' '.join(self.display))
        if self.guessed:
            print(colorize(f"Открытые буквы: {', '.join(sorted(self.guessed))}", 'green'))
        if self.wrong:
            print(colorize(f"Ошибочные: {', '.join(sorted(self.wrong))}", 'red'))

    def guess(self, letter):
        if letter in self.guessed or letter in self.wrong:
            print(colorize("Вы уже называли эту букву.", 'yellow'))
            return False
        if letter in self.word:
            self.guessed.add(letter)
            for i, ch in enumerate(self.word):
                if ch == letter:
                    self.display[i] = ch
            if '_' not in self.display:
                self.won = True
                self.game_over = True
            return True
        else:
            self.wrong.add(letter)
            self.errors += 1
            if self.errors >= self.max_errors:
                self.game_over = True
            return False

    def hint(self):
        if self.hint_used:
            print(colorize("Подсказка уже использована.", 'yellow'))
            return
        # Находим первую неоткрытую букву
        for i, ch in enumerate(self.word):
            if self.display[i] == '_':
                self.display[i] = ch
                self.guessed.add(ch)
                self.hint_used = True
                print(colorize(f"Подсказка: буква '{ch}' открыта.", 'cyan'))
                if '_' not in self.display:
                    self.won = True
                    self.game_over = True
                break

    def play(self):
        print(colorize("Добро пожаловать в Виселицу!", 'bold'))
        print("Вводите буквы. ? - подсказка. quit - выход.")
        while not self.game_over:
            self.display()
            cmd = input("> ").strip().lower()
            if cmd == 'quit':
                print("Выход.")
                return
            if cmd == '?':
                self.hint()
                continue
            if len(cmd) != 1 or not cmd.isalpha():
                print(colorize("Введите одну букву.", 'red'))
                continue
            self.guess(cmd)
        # Конец игры
        self.display()
        if self.won:
            print(colorize("🎉 Поздравляем! Вы выиграли!", 'green'))
            self.stats['wins'] += 1
            self.stats['streak'] += 1
            if self.stats['streak'] > self.stats['best_streak']:
                self.stats['best_streak'] = self.stats['streak']
        else:
            print(colorize(f"💀 Вы проиграли. Загаданное слово: {self.word}", 'red'))
            self.stats['losses'] += 1
            self.stats['streak'] = 0
        self.stats['games'] += 1
        self.save_stats()
        print(colorize(f"Статистика: игр {self.stats['games']}, побед {self.stats['wins']}, поражений {self.stats['losses']}, серия {self.stats['streak']}, лучшая {self.stats['best_streak']}", 'blue'))

def main():
    max_errors = 10
    theme = 'general'
    show_stats = False
    reset_stats = False
    args = sys.argv[1:]
    for i, arg in enumerate(args):
        if arg in ['easy', 'medium', 'hard']:
            if arg == 'easy': max_errors = 10
            elif arg == 'medium': max_errors = 7
            elif arg == 'hard': max_errors = 5
        elif arg == '-t' or arg == '--theme':
            if i+1 < len(args):
                theme = args[i+1]
                if theme not in WORDS:
                    theme = 'general'
        elif arg == '-s' or arg == '--stats':
            show_stats = True
        elif arg == '-r' or arg == '--reset':
            reset_stats = True
        elif arg == '-h' or arg == '--help':
            print("Usage: hangman.py [easy|medium|hard] [-t theme] [-s] [-r]")
            return
    if reset_stats:
        stats_file = Path.home() / '.hangman_stats.json'
        if stats_file.exists():
            stats_file.unlink()
        print("Статистика сброшена.")
        return
    if show_stats:
        stats_file = Path.home() / '.hangman_stats.json'
        if stats_file.exists():
            with open(stats_file, 'r') as f:
                stats = json.load(f)
                print(colorize("📊 Статистика:", 'bold'))
                print(f"  Игр: {stats['games']}")
                print(f"  Побед: {stats['wins']}")
                print(f"  Поражений: {stats['losses']}")
                print(f"  Текущая серия: {stats['streak']}")
                print(f"  Лучшая серия: {stats['best_streak']}")
        else:
            print("Статистика пуста.")
        return
    game = Hangman(max_errors, theme)
    game.play()

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print(colorize("\nИгра прервана.", 'yellow'))
        sys.exit(0)

File: hangman.rb
#!/usr/bin/env ruby
# hangman.rb
# encoding: UTF-8

require 'json'
require 'fileutils'
require 'set'

COLORS = {
  reset: "\e[0m",
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  blue: "\e[94m",
  cyan: "\e[96m",
  bold: "\e[1m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

WORDS = {
  'animals' => %w[кот собака тигр слон жираф дельфин медведь волк лиса заяц],
  'cities' => %w[москва париж лондон берлин рим токио пекин нью-йорк каир мехико],
  'food' => %w[пицца суши бургер паста салат суп стейк омлет торт мороженое],
  'professions' => %w[врач учитель инженер пилот повар художник музыкант архитектор юрист журналист],
  'general' => %w[программирование алгоритм компьютер интернет игра код данные сервер сайт приложение]
}

HANGMAN_PICS = [
  <<~PIC,
      +---+
          |
          |
          |
          |
          |
    =========
  PIC
  <<~PIC,
      +---+
      |   |
          |
          |
          |
          |
    =========
  PIC
  <<~PIC,
      +---+
      |   |
      O   |
          |
          |
          |
    =========
  PIC
  <<~PIC,
      +---+
      |   |
      O   |
      |   |
          |
          |
    =========
  PIC
  <<~PIC,
      +---+
      |   |
      O   |
     /|   |
          |
          |
    =========
  PIC
  <<~PIC,
      +---+
      |   |
      O   |
     /|\\  |
          |
          |
    =========
  PIC
  <<~PIC,
      +---+
      |   |
      O   |
     /|\\  |
     /    |
          |
    =========
  PIC
  <<~PIC,
      +---+
      |   |
      O   |
     /|\\  |
     / \\  |
          |
    =========
  PIC
]

class Hangman
  attr_reader :max_errors, :theme, :word, :display, :guessed, :wrong,
              :errors, :hint_used, :game_over, :won, :stats, :stats_file

  def initialize(max_errors = 10, theme = 'general')
    @max_errors = max_errors
    @theme = theme
    @stats_file = File.join(Dir.home, '.hangman_stats.json')
    load_stats
    words = WORDS[theme] || WORDS['general']
    @word = words.sample
    @display = Array.new(@word.length, '_')
    @guessed = Set.new
    @wrong = Set.new
    @errors = 0
    @hint_used = false
    @game_over = false
    @won = false
  end

  def load_stats
    if File.exist?(@stats_file)
      @stats = JSON.parse(File.read(@stats_file))
    else
      @stats = { 'games' => 0, 'wins' => 0, 'losses' => 0, 'streak' => 0, 'best_streak' => 0 }
    end
  end

  def save_stats
    File.write(@stats_file, JSON.pretty_generate(@stats))
  end

  def display_state
    system('clear') || system('cls')
    puts colorize('🎯 ВИСЕЛИЦА', :bold)
    puts colorize("Тема: #{@theme}", :blue)
    col = @errors > 0 ? :red : :green
    puts colorize("Ошибок: #{@errors}/#{@max_errors}", col)
    puts HANGMAN_PICS[@errors]
    puts "Слово: #{@display.join(' ')}"
    unless @guessed.empty?
      puts colorize("Открытые буквы: #{@guessed.to_a.join(', ')}", :green)
    end
    unless @wrong.empty?
      puts colorize("Ошибочные: #{@wrong.to_a.join(', ')}", :red)
    end
  end

  def guess(letter)
    letter = letter.downcase
    if @guessed.include?(letter) || @wrong.include?(letter)
      puts colorize('Вы уже называли эту букву.', :yellow)
      return false
    end
    if @word.include?(letter)
      @guessed.add(letter)
      @word.chars.each_with_index do |ch, i|
        @display[i] = ch if ch == letter
      end
      if !@display.include?('_')
        @won = true
        @game_over = true
      end
      true
    else
      @wrong.add(letter)
      @errors += 1
      @game_over = true if @errors >= @max_errors
      false
    end
  end

  def hint
    if @hint_used
      puts colorize('Подсказка уже использована.', :yellow)
      return
    end
    @display.each_with_index do |ch, i|
      if ch == '_'
        @display[i] = @word[i]
        @guessed.add(@word[i])
        @hint_used = true
        puts colorize("Подсказка: буква '#{@word[i]}' открыта.", :cyan)
        if !@display.include?('_')
          @won = true
          @game_over = true
        end
        break
      end
    end
  end

  def play
    puts colorize('Добро пожаловать в Виселицу!', :bold)
    puts 'Вводите буквы. ? - подсказка. quit - выход.'
    until @game_over
      display_state
      print '> '
      cmd = gets.chomp.strip
      case cmd
      when 'quit'
        puts 'Выход.'
        return
      when '?'
        hint
        next
      else
        if cmd.length != 1 || !cmd.match?(/[[:alpha:]]/)
          puts colorize('Введите одну букву.', :red)
          next
        end
        guess(cmd)
      end
    end
    display_state
    if @won
      puts colorize('🎉 Поздравляем! Вы выиграли!', :green)
      @stats['wins'] += 1
      @stats['streak'] += 1
      if @stats['streak'] > @stats['best_streak']
        @stats['best_streak'] = @stats['streak']
      end
    else
      puts colorize("💀 Вы проиграли. Загаданное слово: #{@word}", :red)
      @stats['losses'] += 1
      @stats['streak'] = 0
    end
    @stats['games'] += 1
    save_stats
    puts colorize("Статистика: игр #{@stats['games']}, побед #{@stats['wins']}, поражений #{@stats['losses']}, серия #{@stats['streak']}, лучшая #{@stats['best_streak']}", :blue)
  end
end

def main
  max_errors = 10
  theme = 'general'
  show_stats = false
  reset_stats = false
  i = 0
  while i < ARGV.size
    arg = ARGV[i]
    case arg
    when 'easy' then max_errors = 10
    when 'medium' then max_errors = 7
    when 'hard' then max_errors = 5
    when '-t', '--theme'
      theme = ARGV[i+1] if i+1 < ARGV.size
      theme = 'general' unless WORDS.key?(theme)
      i += 1
    when '-s', '--stats' then show_stats = true
    when '-r', '--reset' then reset_stats = true
    when '-h', '--help'
      puts 'Usage: ruby hangman.rb [easy|medium|hard] [-t theme] [-s] [-r]'
      return
    end
    i += 1
  end
  if reset_stats
    f = File.join(Dir.home, '.hangman_stats.json')
    File.delete(f) if File.exist?(f)
    puts 'Статистика сброшена.'
    return
  end
  if show_stats
    f = File.join(Dir.home, '.hangman_stats.json')
    if File.exist?(f)
      stats = JSON.parse(File.read(f))
      puts colorize('📊 Статистика:', :bold)
      puts "  Игр: #{stats['games']}"
      puts "  Побед: #{stats['wins']}"
      puts "  Поражений: #{stats['losses']}"
      puts "  Текущая серия: #{stats['streak']}"
      puts "  Лучшая серия: #{stats['best_streak']}"
    else
      puts 'Статистика пуста.'
    end
    return
  end
  game = Hangman.new(max_errors, theme)
  game.play
end

main if __FILE__ == $0
